src/swappy_allocator.rs

```diff
@@ -0,0 +1,1195 @@
+// Copyright (c) 2017-present, PingCAP, Inc. Licensed under Apache-2.0.
+
+//! # Swappy Allocator
+
+use std::alloc::{AllocError, Allocator, Global, Layout};
+use std::fs::{File, OpenOptions};
+use std::path::{Path, PathBuf};
+use std::ptr::{self, NonNull};
+use std::sync::atomic::{AtomicBool, AtomicU32, AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::vec::Vec;
+
+use log::{error, warn};
+use memmap2::MmapMut;
+use parking_lot::Mutex;
+
+const DEFAULT_PAGE_SIZE: usize = 64 * 1024 * 1024; // 64MB
+
+struct SwappyAllocatorCore<A = Global>
+where
+    A: Allocator,
+{
+    budget: usize,
+    path: PathBuf,
+
+    mem_usage: AtomicUsize,
+    mem_allocator: A,
+
+    maybe_swapped: AtomicBool,
+    page_seq: AtomicU32,
+    pages: Mutex<Vec<Page>>,
+}
+
+/// An [`Allocator`] implementation that has a memory budget and can be swapped
+/// out.
+///
+/// The allocations of its internal metadata are not managed (i.e. allocated via
+/// `std::alloc::Global`). Do NOT use it as the global allocator.
+#[derive(Clone)]
+pub struct SwappyAllocator<A: Allocator>(Arc<SwappyAllocatorCore<A>>);
+
+impl<A: Allocator> SwappyAllocator<A> {
+    pub fn new_over(path: &Path, budget: usize, alloc: A) -> SwappyAllocator<A> {
+        if path.exists() {
+            if let Err(e) = std::fs::remove_dir_all(&path) {
+                error!(
+                    "Failed to clean up old swap directory: {}. \
+                    There might be obsolete swap files left in {}.",
+                    e,
+                    path.display()
+                );
+            }
+        }
+        let core = SwappyAllocatorCore {
+            budget,
+            path: path.into(),
+            mem_usage: AtomicUsize::new(0),
+            mem_allocator: alloc,
+            maybe_swapped: AtomicBool::new(false),
+            page_seq: AtomicU32::new(0),
+            pages: Mutex::new(Vec::new()),
+        };
+        SwappyAllocator(Arc::new(core))
+    }
+
+    #[inline]
+    pub fn memory_usage(&self) -> usize {
+        self.0.mem_usage.load(Ordering::Relaxed)
+    }
+
+    #[inline]
+    fn is_swapped(&self, ptr: NonNull<u8>, exaustive_check: bool) -> bool {
+        // Ordering: `maybe_swapped` must be read after the pointer is available.
+        std::sync::atomic::fence(Ordering::Acquire);
+        self.0.maybe_swapped.load(Ordering::Relaxed)
```

Comment:
It's different. fence is more powerful than one atomic access: An acquire atomic load prevents subsequent R/W from being reordered before this load. An acquire fence, on the other hand, prevents subsequent R/W from being reordered before any load before this fence (in essence, it makes all loads before this fence appear as one big atomic load).

Here we use a fence to prevent `maybe_swapped.load()` being ordered before `ptr.load()` (which happens before the scope of this function).

p.s. this is a good writeup: https://preshing.com/20130922/acquire-and-release-fences/

Reply:
always struggle with memory order🥲